packages/devtools/src/ts/devtoolsPanel/messaging.ts

```diff
@@ -27,80 +43,122 @@ const actionHistory$ = new Observable<ActionHistory>(obs => {
-  startWith([]),
-  shareLatest(),
-);
+}).pipe(startWith([]), shareLatest())
 
-type Action = ActionHistory extends Array<infer R> ? R : never;
+type Action = ActionHistory extends Array<infer R> ? R : never
 
-const tagStateReducer = (state: Record<string, DebugTag> = {}, action: Action): Record<string, DebugTag> => {
-  switch(action.type) {
-    case 'newTag$':
-      return {
-        ...state,
-        [action.payload.id]: {
-          ...action.payload,
-          refs: [],
-          latestValues: {},
-        },
-      };
-    case 'tagRefDetection$':
+const tagStateReducer = (
+  state: DebugTag | null,
+  action: Action,
+): DebugTag | null => {
+  if (action.type === "newTag$") {
+    return {
+      ...action.payload,
+      refs: [],
+      latestValues: {},
+    }
+  }
+  if (state === null) {
+    return null
+  }
+  switch (action.type) {
+    case "tagRefDetection$":
       return {
         ...state,
-        [action.payload.id]: {
-          ...state[action.payload.id],
-          refs: [...state[action.payload.id].refs, action.payload.ref],
-        },
-      };
-    case 'tagSubscription$':
+        refs: [...state.refs, action.payload.ref],
+      }
+    case "tagSubscription$":
       return {
         ...state,
-        [action.payload.id]: {
-          ...state[action.payload.id],
-          latestValues: {
-            ...state[action.payload.id].latestValues,
-            [action.payload.sid]: undefined,
-          },
+        latestValues: {
+          ...state.latestValues,
+          [action.payload.sid]: undefined,
         },
-      };
-    case 'tagUnsubscription$':
-      const { [action.payload.sid]: _, ...latestValues } = state[
-        action.payload.id
-      ].latestValues;
+      }
+    case "tagUnsubscription$":
+      const { [action.payload.sid]: _, ...latestValues } = state.latestValues
       return {
         ...state,
-        [action.payload.id]: {
-          ...state[action.payload.id],
-          latestValues,
-        },
-      };
-    case 'tagValueChange$':
+        latestValues,
+      }
+    case "tagValueChange$":
       const values = {
-        ...state[action.payload.id].latestValues,
-      };
+        ...state.latestValues,
+      }
       if (values[action.payload.sid] === action.payload.value) {
-        return state;
+        return state
       }
-      values[action.payload.sid] = action.payload.value;
+      values[action.payload.sid] = action.payload.value
       return {
         ...state,
-        [action.payload.id]: {
-          ...state[action.payload.id],
-          latestValues: values,
-        },
-      };
+        latestValues: values,
+      }
   }
 }
 
-export const tagValue$ = (index: number | null) => actionHistory$.pipe(
-  map(history => {
-    const slice = index === null ? history : history.slice(0, index);
-    return slice.reduce(tagStateReducer, {})
-  })
-);
+const empty = Symbol("empty")
+const incremental = () => <T>(source: Observable<T[]>) =>
```

Comment:
This operator is interesting: it takes an observable of an array, and it kind-of does the equivalent of `from()`, but only emits the values not emitted before.

This assumes that the previous values don't change, and new values are added at the end (covers our use case). It's resettable: If the next value is an array with smaller length than the previous one, it will emit a "reset" message before sending the whole new array again.

Why do we need this? The history of actions gets accumulated into an array, which can be reduced to the state of each tag, but that history can get sliced anytime if the user wants to see a state in the past.
Moreover, every tag node will listen for any update it's interested in, and it's a way of preventing every tag from having to recompute all the states whenever a new action comes in.